application/builtin/contract/deposit/deposit.go

```diff
@@ -294,48 +300,65 @@ func (d *Deposit) checkConfirm(migrationDaemonRef string, amountStr string) (boo
 	if len(activateDaemons) > 0 {
 		canConfirm, err := d.checkAmount(activateDaemons, migrationDaemonRef, amountStr)
 		if err != nil {
-			return canConfirm, fmt.Errorf("failed to check amount in confirmation from migration daemon: %s", err.Error())
+			return canConfirm, errors.Wrap(err, "failed to check amount in confirmation from migration daemon")
 		}
 		return canConfirm, nil
 	}
 	return false, nil
 }
 
-func (d *Deposit) canTransfer(transferAmount *big.Int) error {
+func (d *Deposit) availableAmount() (*big.Int, error) {
 	if d.VestingType == appfoundation.DefaultVesting && !d.IsConfirmed {
-		return fmt.Errorf("number of confirms is less then 2")
+		return nil, errors.New("number of confirms is less then 2")
 	}
 
 	currentPulse, err := foundation.GetPulseNumber()
 	if err != nil {
-		return fmt.Errorf("failed to get pulse number: %s", err.Error())
+		return nil, errors.Wrap(err, "failed to get pulse number")
 	}
-	if d.PulseDepositUnHold > currentPulse {
-		return fmt.Errorf("hold period didn't end")
+	if currentPulse < d.PulseDepositUnHold {
+		return nil, errors.New("hold period didn't end")
 	}
 
-	spentPeriodInPulses := big.NewInt(int64(currentPulse-d.PulseDepositUnHold) / d.VestingStep)
 	amount, ok := new(big.Int).SetString(d.Amount, 10)
 	if !ok {
-		return fmt.Errorf("can't parse derposit amount")
+		return nil, errors.New("can't parse derposit amount")
 	}
 	balance, ok := new(big.Int).SetString(d.Balance, 10)
 	if !ok {
-		return fmt.Errorf("can't parse derposit balance")
+		return nil, errors.New("can't parse derposit balance")
+	}
+
+	// Allow to transfer whole balance if vesting period has already finished
+	if currentPulse > d.PulseDepositUnHold+insolar.PulseNumber(d.Vesting) {
+		return balance, nil
 	}
 
-	// How much can we transfer for this time
-	availableForNow := new(big.Int).Div(
-		new(big.Int).Mul(amount, spentPeriodInPulses),
-		big.NewInt(d.Vesting/d.VestingStep),
+	// Total number of vesting steps in vesting period
+	totalSteps := big.NewInt(d.Vesting / d.VestingStep)
+	// Vesting steps already passed by now
+	passedSteps := big.NewInt(int64(currentPulse-d.PulseDepositUnHold) / d.VestingStep)
+	// Amount that has been vested by now
+	vestedByNow := new(big.Int).Div(
+		new(big.Int).Mul(amount, passedSteps),
+		totalSteps,
 	)
+	// Amount that is still locked on deposit
+	onHold := new(big.Int).Sub(amount, vestedByNow)
+	// Amount that is now available for withdrawal
+	availableNow := new(big.Int).Sub(balance, onHold)
```

Comment:
this can become negative when balance is 0 and vesting already started

Reply:
Fixed. Returning zero in this case